python/ray/data/_internal/batcher.py

```diff
@@ -316,5 +374,113 @@ def next_batch(self) -> Block:
             self._batch_head : self._batch_head + batch_size
         ]
         self._batch_head += batch_size
+
         # Yield the shuffled batch.
-        return BlockAccessor.for_block(self._shuffle_buffer).take(batch_indices)
+        batch = BlockAccessor.for_block(self._shuffle_buffer).take(batch_indices)
+
+        # Format the batch to the provided batch format.
+        batch = BlockAccessor.for_block(batch).to_batch_format(self._batch_format)
+
+        return batch
+
+
+class AsyncBatcher(BatcherInterface):
```

Comment:
**Design note:** Rather than implementing the `BatcherInterface` with async/background batching being an implementation detail, where the `AsyncBatcher` ends up mostly being a proxy to the base batcher, I'm wondering if it would be cleaner to lift the backgrounding of the batcher out of the abstraction, e.g. where the background thread and queue are defined outside of the batcher, and `batcher.next()` is called in the background thread. This is a common design pattern of separating the operation (in this case batch preparation) from the execution and operator-to-operator communication model (background thread and queue), and should make it easier to change/add more operations and tweak the execution/communication model.

If we refactor the batch producer (`batch_blocks`) code to be a chained sequence of `Iterable[Data] -> Iterable[Data]` generators, I think we can more cleanly separate the data processing code (e.g. prefetching, batching, formatting, etc.) from the execution/orchestration code (serial vs. background thread vs. threadpool vs. actor, and iterator vs. queue); this is similar to the [Torch `IterDataPipe` paradigm](https://github.com/pytorch/pytorch/issues/49440), and they essentially had the same motivations as us. We could then have an `Iterable[ray.ObjectRef[Block]] -> Iterable[Block]` (pre)fetcher, an `Iterable[Block] -> Iterable[Batch]` batcher, and an `Iterable[Batch] -> Iterable[Batch]` formatter.

With --> representing a simple iterable feed (no queueing/buffering), the status quo (no batcher background thread) is
```
producer -(plasma_queue)-> fetcher --> batcher --> formatter --> consumer
```
where it's single-threaded (except for the Ray Core object store producer, which is multithreaded) and serial. This PR would be changing this to
```
producer -(plasma_queue)-> fetcher --> [batcher --> formatter] -(queue)-> consumer
```
where everything in `[]` is in a background thread. Explicitly representing this in the `batch_blocks` code would make it easier to tweak this going forward, e.g. moving the fetcher into the background thread as well, doing some of the `fetcher --> batcher --> formatter` chain in parallel over a threadpool, etc.

What do you think? And is this somewhat what your part 2 PR looks like?

Reply:
This structuring should also be a lot easier to incorporate into the upcoming execution model overhaul.

Reply:
yes this is what part 2 of the pr looks like!

Reply:
Would it be reasonable to bring this part 1 PR closer to that paradigm? Otherwise we're essentially adding an `AsyncBatcher` abstraction (which I think we agree is not the design that we'd like now that we're introducing asynchrony here) in this PR only to remove it in the follow-up stacked PR.

I'm a big fan of the "make the change easy, then make the easy change" principle, and I think that the sequence here would be:
1. Refactor to the chained generators pattern.
2. Push batching and formatting into a background thread, with formatted batches pushed into a queue.
3. Unify the block prefetching with batch prefetching.

Where (1) and (2) can probably be the same PR (i.e. this PR) if the diff is small enough and is pretty easy to review.

Reply:
Even in the followup there is an AsyncBatcher abstraction, but the dataflow pattern is still Iterable based.

producer -(plasma_queue)-> fetcher --> batcher --> formatter --> consumer

Basically, AsyncBatcher abstracts the batcher and formatter part of this flow.

Reply:
are you recommending to remove the Batcher concept altogether?

Reply:
Sure we can go with the 3 step approach that you mentioned...

but the Batcher class that does `Iterable[Block] -> Iterable[Batch]` would still be useful so that we don't have a bunch of loose generator functions hanging around.

Reply:
Ah, so I'm contending that the `AsyncBatcher` abstraction is probably not the design we want since it's a tight coupling of the execution model + the queueing logic between the operators with the implementation of a single (batcher + formatter) operator. That coupling will make it difficult to change or add more operators and tweak the execution/queueing models, to unit test these different pockets of logic, and to reason through the execution/queueing model.